### DISCUSSION TITLE
I finally got a Heltec CubeCell-GPS HTCC-AB02s to be a Lora Sensor on the FDRS network

### DISCUSSION THREAD

**Discussion by sbobowski** [timmbogner/Farm-Data-Relay-System#171] [Show and tell]
Sorry, just so happy to have figured this out.
Forewarning: I only got the basics down on this.

After Googling every last possible search string I could come up with, enough breadcrumbs, rewriting a portion of the fdrs node code, rewriting part of the RadioLib code (throwing both rewrites away once I discovered the needed work was already complete) and a little debugging got me to where I am. (sorry if someone else covered this before, sorry, I could not find it, I've been at this for a week).

Let me walk through the code with you (be gentle, I'm not a pro).
I started with the LoRa Sensor Example in the FDRS library

Make sure you set your secrets in the **fdrs_globals.h** file.

**LoRa_Sensor.ino:** (no changes here)
```
//  FARM DATA RELAY SYSTEM
//
//  LoRa Sensor Example
//
//  Developed by Timm Bogner (timmbogner@gmail.com) in Urbana, Illinois, USA.
//  An example of how to send data via LoRa using FDRS.
//

#include "fdrs_node_config.h"
#include <fdrs_node.h>

float data1;
float data2;

void setup() {
  beginFDRS();
}
void loop() {
  data1 = readHum();
  loadFDRS(data1, HUMIDITY_T);
  data2 = readTemp();
  loadFDRS(data2, TEMP_T);
  sendFDRS();
  sleepFDRS(10);  //Sleep time in seconds
}

float readTemp() {
  return 21.0;
}

float readHum() {
  return random(0,100);
}
```

**fdrs_node_config.h:**
- updated my READING_ID for this sensor
- updated my GTWY_MAC to match my LoRa board being used as a receiver (I'm using a Heltec ESP32 V3 as a gateway, not sure if I'm allowed to link to Amazon for the bards I used)
- set the RADIOLIB_MOD to SX1262
- set the LORA pins from the Heltec callout for the AB02s
- declared the USE_SX126X

```
//  FARM DATA RELAY SYSTEM
//
//  Sensor Configuration

#include <fdrs_globals.h>

#define READING_ID    3   //Unique ID for this sensor
#define GTWY_MAC      0x01 //Address of the nearest gateway

//#define USE_ESPNOW
#define USE_LORA
#define DEEP_SLEEP
//#define POWER_CTRL    14
#define FDRS_DEBUG

// LoRa Configuration
#define RADIOLIB_MODULE SX1262
#define LORA_SS    RADIO_NSS
#define LORA_RST   RADIO_RESET
#define LORA_DIO   RADIO_DIO_1
#define LORA_BUSY  RADIO_BUSY
#define USE_SX126X

#define LORA_TXPWR 17   // LoRa TX power in dBm (: +2dBm - +17dBm (for SX1276-7) +20dBm (for SX1278))
#define LORA_ACK        // Request LoRa acknowledgment.

//#define CUSTOM_SPI
//#define LORA_SPI_SCK 5
//#define LORA_SPI_MISO 19
//#define LORA_SPI_MOSI 27
```

The big holdup was that there is no "yield" function for the CubeCell-GPS board.
so I modified the FDRS library to exclude it for the one board type where yield is called in the pingFDRSLoRa function.
**fdrs_node_lora.h**
- I wrapped the yield(); call on line 497 in a not defined wrapper
```
// FDRS Sensor pings gateway and listens for a defined amount of time for a reply
// Blocking function for timeout amount of time (up to timeout time waiting for reply)(IE no callback)
// Returns the amount of time in ms that the ping takes or predefined value if ping fails within timeout
uint32_t pingFDRSLoRa(uint16_t *address, uint32_t timeout)
{
    SystemPacket sys_packet = {.cmd = cmd_ping, .param = 0};

    transmitLoRa(address, &sys_packet, 1);
    DBG("LoRa ping sent to address: 0x" + String(*address, HEX));
    uint32_t ping_start = millis();
    pingFlag = false;
    while ((millis() - ping_start) <= timeout)
    {
        handleLoRa();
		#ifndef CubeCell_GPS
        yield(); // do I need to yield or does it automatically?
		#endif
        if (pingFlag)
        {
            DBG("LoRa Ping Returned: " + String(millis() - ping_start) + "ms.");
            pingFlag = false;
            return (millis() - ping_start);
        }
    }
    DBG("No LoRa ping returned within " + String(timeout) + "ms.");
    return UINT32_MAX;
}
```

And that's it!!! really it is just modifying the library and it is no different than configuring the ESP32 V3 LoRa board.
Hope that helps somebody (who might be in the same boat as me with ~20 of these HTCC-AB02s boards laying around).

An epic shoutout to @timmbogner for creating this library!!! it is everything I wanted to do on my farm with the LoRa boards but never sat long enough to standardize a communication standard (and you included ESPNow, which eventually I'm going to take a stab at with the library for sending ESP-Cam over ESPNow)

**Comment by timmbogner** [comment]
That's fantastic info, thanks for working through it and testing! It's funny because @aviateur17 and I have been discussing yield() usage recently. I think that ```yield();``` call can actually be wrapped with ```#ifdef ESP8266``` because it is not needed by any other MCUs that I know of. If you're up for it you can submit this change as a pull request, otherwise I can make the change and give you credit as a co-author on the commit.

I'd be quite interested in whether the gateway works on CubeCell too. There are no yields, so it may work out the box.

Thanks a lot! It actually started with ESP-Now, but I snuck LoRa in after some sagely advice from a cool Swiss guy.

I do hope to one day handle a new 'bulk' datatype in FDRS to allow longer, framed transmissions (ex: an image from ESP-CAM). There are a lot of features to add before that, though!

**Reply by sbobowski** [reply]
@timmbogner Thank you so much for allowing me to submit this as a change request (I think I submitted it correctly) This is a huge deal to me.
The cool Swiss guy is the one that turned me onto using LoRa without LoRaWAN.
I can't wait to see what you and the other maintainers come up with. And thank you again for creating this library!!!!
